<center><h3><b>Optimização Heurística</b></h3></center>
<center><h3><b>2021/2022</b></h3></center>
<br>
<center><h3><b>Trabalho de Grupo</b></h3></center>


<div class="alert alert-block alert-info">
    <b>Trabalho Realizado por:</b>
    <br>
    <br>
    <li>André Novo nº93343</li>
      <li>Luís Pereira nº98398</li>
      <li>Mariana Semedo nº98322</li>
      <li>Maria Pais nº98263</li>
      <li>Sebastião Rosalino nº98437</li>
      <li>Tiago Pratas nº98952</li></div>
<br>

#### Bibliotecas utilizadas

In [1]:
import pandas as pd
import numpy as np
from itertools import permutations
import random
from itertools import combinations

|Regiões|R1|R2|R3|R4|R5|R6|R7|R8|R9|R10|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|**Procura, em unidades, da Região i**|80|270|250|160|180|230|120|90|150|210|

||R1|R2|R3|R4|R5|R6|R7|R8|R9|R10|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|**L1**|4|5|6|8|10|12|7|9|7|9|
|**L2**|6|4|3|5|8|5|10|5|6|2|
|**L3**|9|7|4|3|4|9|5|7|8|14|
|**L4**|12|3|9|8|6|4|12|10|4|3|
|**L5**|3|13|7|4|3|10|6|4|15|7|

|Locais para instalar armazéns|L1|L2|L3|L4|L5|
|:---:|:---:|:---:|:---:|:---:|:---:|
|**Custo fixo, em u.m, de instalação de um armazém no local i**|1000|1000|1000|1000|1000|
|**Capacidade, em unidades, do armazem instalado no local i**|900|900|900|900|900|

<div class="alert alert-block alert-info">
    <b>a) Descreva, por palavras, uma solução admissível para o problema da empresa Lusa
    </b>
<br>

Uma solução admissível para o problema será aquela que satisfaça as procuras das regiões, respeite a capacidade dos armazéns instalados e minimize o custo total.

Soma das Procuras = 1740

Dois armazéns podem chegar pois conseguem armazenar 1800 unidades do produto.

Custos totais:

* Armazém instalado no local 1 = 77

* Armazém instalado no local 2 = 54

* Armazém instalado no local 3 = 70

* Armazém instalado no local 4 = 71

* Armazém instalado no local 5 = 72

Com base nesta heurística escolhemos os locais 2 e 3, cujos custos totais são os menores.

Os armazéns vão ser instalados nos locais L2 e L3.

Com base nos custos de transporte entre cada local e cada região, escolheram-se as ligações de transporte com base nos custos mínimos, sem exceder a capacidade de cada armazém.

* A procura na R1 vai ser garantida por L2 (900 - 80) ou seja terá nova capacidade de 820

* A procura na R2 vai ser garantida por L2 (820 - 270) ou seja terá nova capacidade de 550

* A procura na R3 vai ser garantida por L2 (550 - 250) ou seja nova capacidade de 300

* A procura na R4 vai ser garantida por L3 (900 - 160) ou seja nova capacidade de 740

* A procura na R5 vai ser garantida por L3  (740 - 180) ou seja nova capacidade de 560

* A procura na R6 vai ser garantida por L2  (300 - 230) ou seja nova capacidade de 70

* A procura na R7 vai ser garantida por L3  (560 - 120) ou seja nova capacidade de 440

* A procura na R8 vai ser garantida por L3  (440 - 90) ou seja nova capacidade de 350

* A procura na R9 vai ser garantida por L3  (350 - 150) ou seja nova capacidade de 200

Como já nem L2 nem L3 conseguem abastecer R10, construir-se-á um terceiro um terceiro armazém no local mais próximo de R10, ou seja, L4.

* A procura em R10 vai ser garantida por L4 (900 - 210) ou seja nova capacidade de 690.

O custo total associado a esta solução é:

3 * Custos fixos de instalação de armazéms + 80 * 6 + 4 * 270 + 3 * 250 + 3 * 160 + 4 * 180 + 5 * 230 + 5 * 120 + 7 * 90 + 8 * 150 + 3 * 210

|Região| Local de instalação | Custo unitário | Nº de unidades|
|:---:|:---:|:---:|:---:|
|**R1**|L2|6|80|
|**R2**|L2|4|270|
|**R3**|L2|3|250|
|**R4**|L3|3|160|
|**R5**|L3|4|180|
|**R6**|L2|5|230|
|**R7**|L3|5|120|
|**R8**|L3|7|90|
|**R9**|L3|8|150|
|**R10**|L4|3|210|

In [2]:
dict1 = {"custo":[6,4,3,3,4,5,5,7,8,3],"n_units":[80,270,250,160,180,230,120,90,150,210]}
table1 = pd.DataFrame(dict1)
n_locals = 3
cost_final = sum(table1["custo"]*table1["n_units"]) + n_locals*1000
cost_final

10720

O custo total foram 10720 unidades monetárias.

Esta solução consiste instalar os armazéns nas regiões L2, L3 e L4, com um custo de 10720.

<div class="alert alert-block alert-info">
    <b>b) Suponha que foram instalados armazéns nos locais L1 e L3. Desenvolva uma heurística para determinar as unidades de produto que devem ser transportadas a partir dos locais L1 e L3 para cada uma das regiões de forma a satisfazer as suas procuras e a respeitar as capacidades dos armazéns. Com base na heurística desenvolvida, apresente uma solução admissível para o problema da empresa Lusa.
    </b>
<br>



Vamos instalar os armazéns nos locais L1 e L3.

O custo de instalação é 2000.

A heurística definida consiste em começar a abastecer as regiões em ordem decrescente de procura, abastacendo as mesmas com o local que apresentar o custo unitário mínimo (de forma a minimizar os custos de transporte), respeitando as capacidades dos armazéns.

Passando aos abastecimentos:

* A procura na R2 vai ser garantida por L1 (900 - 270) ou seja nova capacidade de 630

* A procura na R3 vai ser garantida por L3 (900 - 250) ou seja nova capacidade de 650

* A procura na R6 vai ser garantida por L3 (650 - 230) ou seja nova capacidade de 420

* A procura na R10 vai ser garantida por L1 (630 - 210) ou seja nova capacidade de 420

* A procura na R5 vai ser garantida por L3 (420 - 180) ou seja nova capacidade de 240

* A procura na R4 vai ser garantida por L3 (240 - 160) ou seja nova capacidade de 80

* A procura na R9 vai ser garantida por L1 (420 - 150) ou seja nova capacidade de 270

* A procura na R7 vai ser garantida por L1, **por falta de capacidade em L3** (270 - 120) ou seja nova capacidade de 150

* A procura na R8 vai ser garantida por L1, **por falta de capacidade em L3** (150 - 90) ou seja nova capacidade de 60

* A procura na R1 vai ser garantida por L3, **por falta de capacidade em L1** (80 - 80) ou seja capacidade de 0

O custo total desta solução é:

2 * Custos fixos de instalação dos armazéns + 5 * 270 + 4 * 250 + 9 * 230 + 9 * 210 + 4 * 180 + 3 * 160 + 7 * 150 + 7 * 120 + 9 * 90 + 9 * 80

|Região| Local de instalação | Custo unitário | Nº de unidades|
|:---:|:---:|:---:|:---:|
|**R1**|L3|9|80|
|**R2**|L1|5|270|
|**R3**|L3|4|250|
|**R4**|L3|3|160|
|**R5**|L3|4|180|
|**R6**|L3|9|230|
|**R7**|L1|7|120|
|**R8**|L1|9|90|
|**R9**|L1|7|150|
|**R10**|L1|9|210|

In [2]:
import pandas as pd
dict2 = {"custo":[9,5,4,3,4,9,7,9,7,9],"n_units":[80,270,250,160,180,230,120,90,150,210]}
table2 = pd.DataFrame(dict2)
n_locals = 2
cost_final = sum(table2["custo"]*table2["n_units"]) + n_locals*1000
cost_final

12930

<div class="alert alert-block alert-info">
    <b>h) Com base nas alíneas anteriores, implemente um algoritmo genético que permita gerar uma
solução admissível de qualidade para o problema da empresa Lusa. Dê uma breve explicação
do algoritmo genético que implementou, nomeadamente dimensão da população, forma como
é gerada a população inicial, método de seleção, método de substituição da população e
critério(s) de paragem.
    </b>
<br>

### Algoritmo genético

* Operador de mutação ***Bit Flip***

In [3]:
def mutation(x):
    p_mutacao = 0.1
    probs = np.random.rand(len(x))
    for i in range(len(x)):
        if probs[i]<= p_mutacao:
            x[i]=x[i]^1
    if sum(x)<2:
        mutation(x)
    return x

* Definição das matrizes de **custos de transporte** e de **procura por região** e ainda da capacidade máxima por armazém

In [4]:
# matriz de custos de transporte
distance_matrix = pd.DataFrame([[4, 5, 6, 8, 10, 12, 7, 9, 7, 9],
                                [6, 4, 3, 5, 8, 5, 10, 5, 6, 2],
                                [9, 7, 4, 3, 4, 9, 5, 7, 8, 14],
                                [12, 3, 9, 8, 6, 4, 12, 10, 4, 3],
                                [3, 13, 7, 4, 3, 10, 6, 4, 15, 7]], columns=[0,1, 2, 3, 4, 5, 6, 7, 8, 9])

distance_matrix=distance_matrix.transpose()

# matriz das procuras por região
capacities = pd.Series([80, 270, 250, 160, 180,230,120,90,150,210])

# capacidade máxima por armazém
max_deposit_capacity = 900

* **Definição da heurística definida em b)**

A heurística definida consiste em começar a garantir a satisfação da procura nas regiões por ordem decrescente dessa mesma procura, selecionando para as mesmas o local que apresentar o custo de transporte unitário mínimo (de forma a minimizar os custos de transporte), respeitando as capacidades dos armazéns.

In [5]:
def find_solution(locais):
    a = []
    for n, val in enumerate(locais):
        globals()["local_%d"%n] = val
        dict1 = [globals()["local_%d"%n],capacities[globals()["local_%d"%n]],[]]
        a.append(dict1)
    
    # Matriz dos locais de armazém, capacidade utilizada em cada armazém e quais as regiões nas quais a procura é garantida por 
    # esse mesmo armazém
    deposits = pd.DataFrame(a,columns=['local_armazem', 'used_capacity', 'regioes_servidas'])

    deposits.loc[deposits["local_armazem"]==local_0,"used_capacity"]= deposits.loc[deposits["local_armazem"]==local_0,"used_capacity"]-capacities[local_0]
    deposits.loc[deposits["local_armazem"]==local_1,"used_capacity"]= deposits.loc[deposits["local_armazem"]==local_1,"used_capacity"]-capacities[local_1]
    
    other_cities = pd.Series([1,2,5,9,4,3,8,6,7,0], name='regioes')

    # atribuição da região e respetiva procura ao local com custo de transporte mínimo 
    transport_cost=0
    for j, city in other_cities.items():
        closest_deposit = distance_matrix.loc[city, deposits['local_armazem']].idxmin()
        if closest_deposit == local_0:
            other_deposit = local_1
        else:
            other_deposit = local_0
        if deposits.loc[deposits['local_armazem'] == closest_deposit, 'used_capacity'].values[0] + capacities[
            city] <= max_deposit_capacity:
            deposits.loc[deposits['local_armazem'] == closest_deposit, 'regioes_servidas'].values[0].append(city+1)
            deposits.loc[deposits['local_armazem'] == closest_deposit, 'used_capacity'] += capacities[city]
            
            # soma total dos custos de transporte
            transport_cost += distance_matrix.loc[city, closest_deposit]*capacities[city]
        else:
            deposits.loc[~deposits['local_armazem'].isin([closest_deposit]), 'regioes_servidas'].values[0].append(city+1)
            deposits.loc[~deposits['local_armazem'].isin([closest_deposit]), 'used_capacity'] += capacities[city]
            
            # soma total dos custos de transporte
            transport_cost += distance_matrix.loc[city, other_deposit]*capacities[city]
            
    deposits["local_armazem"] += 1
    
    # soma do custo de instalação dos n armazéns
    transport_cost += (1000*len(locais))

    return int(transport_cost)



* Definição da **população inicial** e custos associados/valores de aptidão

In [7]:
locais = list([1,0,1,0,0])
permutation_list = list(permutations(locais))
df_permutations = pd.DataFrame(permutation_list, columns=locais)
df_permutations = df_permutations.drop_duplicates()
df_inicial = df_permutations.copy()
df_inicial.columns = [1,2,3,4,5]
custos = [df_inicial.columns[i] for j in range(0, len(df_inicial)) for i in range(0, len(df_inicial.columns)) if df_inicial.iloc[j,i]==1]
split_custos = [custos[x:x+2] for x in range(0,len(custos),2)]

for i in range(0,len(split_custos)):
    for j in range(0,2):
        split_custos[i][j]= split_custos[i][j]-1

total = [find_solution(i) for i in split_custos]
print(total)
    
df_inicial["Custo"]=total
print(df_inicial)

[12930, 11240, 12700, 12820, 11970, 9990, 9850, 9110, 12130, 9210]
    1  2  3  4  5  Custo
0   1  0  1  0  0  12930
2   1  0  0  1  0  11240
3   1  0  0  0  1  12700
6   1  1  0  0  0  12820
24  0  1  1  0  0  11970
26  0  1  0  1  0   9990
27  0  1  0  0  1   9850
36  0  0  1  1  0   9110
37  0  0  1  0  1  12130
40  0  0  0  1  1   9210


### Aplicação do algoritmo genético definido

<div class="alert alert-warning" role="alert">
    <b>AVISO</b>
</div>

Custos correspondentes a NaN indicam soluções não admissíveis cuja reparação de admissibilidade se dará posteriormente quando aplicado o operador de mutação.

In [8]:
import warnings
warnings.filterwarnings("ignore")

# Definição do nº inicial de iterações
n_iters = 8 
i = 0

while i < n_iters:
    i += 1
    # k representa o número de cromossomas selecionados da população de forma aleatória

    # Seleção do Pai1: dos k cromossomas selecionados, o Pai1 corresponde ao cromossoma de melhor aptidão (i.e., de menor custo)
    pai_list1 = random.sample(list(df_inicial.index), 3)
    pai1 = df_inicial.loc[pai_list1, 'Custo'].idxmin()

    # Seleção do Pai2: dos k cromossomas selecionados, o Pai2 corresponde ao cromossoma de melhor aptidão, mas garantindo que é 
    # diferente do Pai1
    abc = True
    while abc:
        pai_list2 = random.sample(list(df_inicial.index), 3)
        pai2 = df_inicial.loc[pai_list2, 'Custo'].idxmin()
        if pai1 != pai2:
            abc = False
    
    # apresentação dos cromossomas pais escolhidos
    print("--------------------------------")
    print("                 Pais           ")
    print('Pai1:',pai1) 
    print('Pai2:',pai2)  
    print("--------------------------------")
    max_distance = distance_matrix.max().max()

    # Geração aleatória do número m para a aplicação do operador crossover a um ponto
    crossover_options = list(range(1, len(locais) - 1))
    m = random.choice(crossover_options)

    df_crossover = df_inicial.loc[[pai1, pai2], :]
    print("     Matriz após crossover      ")
    print(df_crossover)
    print("--------------------------------")
    
    # Geração dos cromossomas filho
    df_crossover.loc['filho1', list(range(1,m+1))] = df_crossover.loc[pai1, list(range(1,m+1))]
    df_crossover.loc['filho1', list(range(m + 1,len(locais)+1))] = df_crossover.loc[pai2, list(range(m + 1,len(locais)+1))]

    #:m
    df_crossover.loc['filho2', list(range(1,m+1))] = df_crossover.loc[pai2, list(range(1,m+1))]
    #m + 1:
    df_crossover.loc['filho2',list(range(m + 1,len(locais)+1))] = df_crossover.loc[pai1, list(range(m + 1,len(locais)+1))]
    
    a = []
    for j in range(0, df_crossover.shape[0]):
        lista = df_crossover.columns[(df_crossover == 1).iloc[j]].values.tolist()
        lista = [x-1 for x in lista]
        a.append(lista)
        
    # Cálculo dos custos associados a cada cromossoma filho    
    if len(a[2])>=2:
        df_crossover.loc["filho1","Custo"]=find_solution(a[2])
    if len(a[3])>=2:
        df_crossover.loc["filho2","Custo"]=find_solution(a[3])

    print(df_crossover)
    print("--------------------------------")
    # Apresentação dos cromossomas filho
    print('Cromossoma Filho1(após crossover):')
    print(df_crossover.loc['filho1',:])
    print()
    print("--------------------------------")
    print('Cromossoma Filho2(após crossover):')
    print(df_crossover.loc['filho2',:])
    print()
    print("--------------------------------")
    
    df_nextgen = df_crossover.copy()
    df_nextgen[[1,2,3,4,5]]  = df_nextgen[[1,2,3,4,5]].astype("int")
    
    # aplicação do operador de mutação bit flip
    import math
    for k in df_nextgen.index:
        if math.isnan(df_nextgen.loc[k,"Custo"]):
            df_nextgen.loc[k,[1,2,3,4,5]]= mutation(list(df_nextgen.loc[k,[1,2,3,4,5]].astype("int")))

    b = []
    for k in range(0, df_nextgen.shape[0]):
        lista = df_nextgen.columns[(df_nextgen == 1.0).iloc[k]].values.tolist()
        lista = [x-1 for x in lista]
        b.append(lista)
    
    # Cálculo dos custos associados aos cromossomas após mutação
    df_nextgen.loc["filho1","Custo"]=find_solution(b[2])
    df_nextgen.loc["filho2","Custo"]=find_solution(b[3])
    
    print(" Matriz de cromossomas após mutações")
    print(df_nextgen)
    print("--------------------------------")
    
    var = df_inicial["Custo"].nlargest(2).index
    df_inicial.loc[var[0],:]= df_nextgen.loc["filho1"]
    df_inicial.loc[var[1],:]= df_nextgen.loc["filho2"]
    print("Substituição dos filhos na população inicial")
    print(df_inicial)



--------------------------------
                 Pais           
Pai1: 36
Pai2: 24
--------------------------------
     Matriz após crossover      
    1  2  3  4  5  Custo
36  0  0  1  1  0   9110
24  0  1  1  0  0  11970
--------------------------------
          1    2    3    4    5    Custo
36      0.0  0.0  1.0  1.0  0.0   9110.0
24      0.0  1.0  1.0  0.0  0.0  11970.0
filho1  0.0  0.0  1.0  0.0  0.0      NaN
filho2  0.0  1.0  1.0  1.0  0.0   9230.0
--------------------------------
Cromossoma Filho1(após crossover):
1        0.0
2        0.0
3        1.0
4        0.0
5        0.0
Custo    NaN
Name: filho1, dtype: float64

--------------------------------
Cromossoma Filho2(após crossover):
1           0.0
2           1.0
3           1.0
4           1.0
5           0.0
Custo    9230.0
Name: filho2, dtype: float64

--------------------------------
 Matriz de cromossomas após mutações
        1  2  3  4  5    Custo
36      0  0  1  1  0   9110.0
24      0  1  1  0  0  11970.0
filh

filho2  0.0  1.0  0.0  1.0  1.0  9000.0
--------------------------------
Cromossoma Filho1(após crossover):
1           0.0
2           1.0
3           0.0
4           1.0
5           1.0
Custo    9000.0
Name: filho1, dtype: float64

--------------------------------
Cromossoma Filho2(após crossover):
1           0.0
2           1.0
3           0.0
4           1.0
5           1.0
Custo    9000.0
Name: filho2, dtype: float64

--------------------------------
 Matriz de cromossomas após mutações
        1  2  3  4  5   Custo
26      0  1  0  1  1  9000.0
0       0  1  0  1  1  9000.0
filho1  0  1  0  1  1  9000.0
filho2  0  1  0  1  1  9000.0
--------------------------------
Substituição dos filhos na população inicial
    1  2  3  4  5  Custo
0   0  1  0  1  1   9000
2   0  1  0  1  1   9000
3   0  1  0  1  1   9000
6   0  1  0  1  1   9000
24  0  1  0  1  1   9000
26  0  1  0  1  1   9000
27  0  1  0  1  1   9000
36  0  1  0  1  1   9000
37  0  1  0  1  1   9000
40  0  1  0  1  1   9000

A solução com menor custo associado é aquela cujos locais de instalação dos armazéns são 2,4 e 5.

### Apresentação das regiões servidas por cada armazém da solução final

Uma vez que a primeira solução apenas retorna o custo total da solução para fins de cálculo do mesmo e aplicação a nível do algoritmo genético, implementa-se a função novamente desta vez a retornar a distribuição das regiões pelos locais de armazenamento.

In [13]:
def find_solution2(locais):
    a = []
    for n, val in enumerate(locais):
        globals()["local_%d"%n] = val
        dict1 = [globals()["local_%d"%n],capacities[globals()["local_%d"%n]],[]]
        a.append(dict1)
    
    # Matriz dos locais de armazém, capacidade utilizada em cada armazém e quais as regiões nas quais a procura é garantida por 
    # esse mesmo armazém
    deposits = pd.DataFrame(a,columns=['local_armazem', 'used_capacity', 'regioes_servidas'])

    deposits.loc[deposits["local_armazem"]==local_0,"used_capacity"]= deposits.loc[deposits["local_armazem"]==local_0,"used_capacity"]-capacities[local_0]
    deposits.loc[deposits["local_armazem"]==local_1,"used_capacity"]= deposits.loc[deposits["local_armazem"]==local_1,"used_capacity"]-capacities[local_1]
    
    other_cities = pd.Series([1,2,5,9,4,3,8,6,7,0], name='regioes')

    # atribuição da região e respetiva procura ao local com custo de transporte mínimo 
    transport_cost=0
    for j, city in other_cities.items():
        closest_deposit = distance_matrix.loc[city, deposits['local_armazem']].idxmin()
        if closest_deposit == local_0:
            other_deposit = local_1
        else:
            other_deposit = local_0
        if deposits.loc[deposits['local_armazem'] == closest_deposit, 'used_capacity'].values[0] + capacities[
            city] <= max_deposit_capacity:
            deposits.loc[deposits['local_armazem'] == closest_deposit, 'regioes_servidas'].values[0].append(city+1)
            deposits.loc[deposits['local_armazem'] == closest_deposit, 'used_capacity'] += capacities[city]
            
            # soma total dos custos de transporte
            transport_cost += distance_matrix.loc[city, closest_deposit]*capacities[city]
        else:
            deposits.loc[~deposits['local_armazem'].isin([closest_deposit]), 'regioes_servidas'].values[0].append(city+1)
            deposits.loc[~deposits['local_armazem'].isin([closest_deposit]), 'used_capacity'] += capacities[city]
            
            # soma total dos custos de transporte
            transport_cost += distance_matrix.loc[city, other_deposit]*capacities[city]
            
    deposits["local_armazem"] += 1
    
    # soma do custo de instalação dos n armazéns
    transport_cost += (1000*len(locais))
    print(deposits)
    print("Custo:",transport_cost)

In [14]:
find_solution2([1,3,4])

   local_armazem  used_capacity regioes_servidas
0              2            460          [3, 10]
1              4            650        [2, 6, 9]
2              5            810  [5, 4, 7, 8, 1]
Custo: 9000


### Aplicação do algoritmo genético para uma população inicial distinta

In [23]:
# matriz de custos de transporte
distance_matrix = pd.DataFrame([[4, 5, 6, 8, 10, 12, 7, 9, 7, 9],
                                [6, 4, 3, 5, 8, 5, 10, 5, 6, 2],
                                [9, 7, 4, 3, 4, 9, 5, 7, 8, 14],
                                [12, 3, 9, 8, 6, 4, 12, 10, 4, 3],
                                [3, 13, 7, 4, 3, 10, 6, 4, 15, 7]], columns=[0,1, 2, 3, 4, 5, 6, 7, 8, 9])

distance_matrix=distance_matrix.transpose()

# matriz das procuras por região
capacities = pd.Series([80, 270, 250, 160, 180,230,120,90,150,210])

# capacidade máxima por armazém
max_deposit_capacity = 900

locall = list([1,2,3,4,5])
locais2 = list([1,1,1,0,0])
permutation_list2 = list(permutations(locais2))
df_permutations2 = pd.DataFrame(permutation_list2, columns=locall)
df_permutations2 = df_permutations2.drop_duplicates()

df_inicial2 = df_permutations2.copy()

listaa = []
for i in range(0, df_inicial2.shape[0]):
    lista = df_inicial2.columns[(df_inicial2 == 1).iloc[i]].values.tolist()
    lista = [x-1 for x in lista]
    listaa.append(lista)
    
total2 = [find_solution(i) for i in listaa]
print(total2)
    
df_inicial2["Custo"]=total2
print('População Inicial')
print(df_inicial2)

[11940, 10440, 11800, 10340, 12160, 9960, 9230, 12040, 9000, 9180]
População Inicial
    1  2  3  4  5  Custo
0   1  1  1  0  0  11940
2   1  1  0  1  0  10440
3   1  1  0  0  1  11800
12  1  0  1  1  0  10340
13  1  0  1  0  1  12160
16  1  0  0  1  1   9960
72  0  1  1  1  0   9230
73  0  1  1  0  1  12040
76  0  1  0  1  1   9000
90  0  0  1  1  1   9180


In [24]:
import warnings
warnings.filterwarnings("ignore")

# Definição do nº inicial de iterações
n_iters = 8
i = 0

while i < n_iters:
    i += 1
    # k representa o número de cromossomas selecionados da população de forma aleatória

    # Seleção do Pai1: dos k cromossomas selecionados, o Pai1 corresponde ao cromossoma de melhor aptidão (i.e., de menor custo)
    pai_list1 = random.sample(list(df_inicial2.index), 3)
    pai1 = df_inicial2.loc[pai_list1, 'Custo'].idxmin()

    # Seleção do Pai2: dos k cromossomas selecionados, o Pai2 corresponde ao cromossoma de melhor aptidão, mas garantindo que é 
    # diferente do Pai1
    abc = True
    while abc:
        pai_list2 = random.sample(list(df_inicial2.index), 3)
        pai2 = df_inicial2.loc[pai_list2, 'Custo'].idxmin()
        if pai1 != pai2:
            abc = False
    
    # apresentação dos cromossomas pais escolhidos
    print("--------------------------------")
    print("                 Pais           ")
    print('Pai1:',pai1) 
    print('Pai2:',pai2)  
    print("--------------------------------")
    max_distance = distance_matrix.max().max()

    # Geração aleatória do número m para a aplicação do operador crossover a um ponto
    crossover_options = list(range(1, len(locais2) - 1))
    m = random.choice(crossover_options)

    df_crossover = df_inicial2.loc[[pai1, pai2], :]
    print("     Matriz após crossover      ")
    print(df_crossover)
    print("--------------------------------")
    
    # Geração dos cromossomas filho
    df_crossover.loc['filho1', list(range(1,m+1))] = df_crossover.loc[pai1, list(range(1,m+1))]
    df_crossover.loc['filho1', list(range(m + 1,len(locais2)+1))] = df_crossover.loc[pai2, list(range(m + 1,len(locais2)+1))]

    #:m
    df_crossover.loc['filho2', list(range(1,m+1))] = df_crossover.loc[pai2, list(range(1,m+1))]
    #m + 1:
    df_crossover.loc['filho2',list(range(m + 1,len(locais2)+1))] = df_crossover.loc[pai1, list(range(m + 1,len(locais2)+1))]
    
    a = []
    for j in range(0, df_crossover.shape[0]):
        lista = df_crossover.columns[(df_crossover == 1).iloc[j]].values.tolist()
        lista = [x-1 for x in lista]
        a.append(lista)
        
    # Cálculo dos custos associados a cada cromossoma filho    
    if len(a[2])>=2:
        df_crossover.loc["filho1","Custo"]=find_solution(a[2])
    if len(a[3])>=2:
        df_crossover.loc["filho2","Custo"]=find_solution(a[3])

    print(df_crossover)
    print("--------------------------------")
    # Apresentação dos cromossomas filho
    print('Cromossoma Filho1(após crossover):')
    print(df_crossover.loc['filho1',:])
    print()
    print("--------------------------------")
    print('Cromossoma Filho2(após crossover):')
    print(df_crossover.loc['filho2',:])
    print()
    print("--------------------------------")
    
    df_nextgen = df_crossover.copy()
    df_nextgen[[1,2,3,4,5]]  = df_nextgen[[1,2,3,4,5]].astype("int")
    
    # aplicação do operador de mutação bit flip
    import math
    for k in df_nextgen.index:
        if math.isnan(df_nextgen.loc[k,"Custo"]):
            df_nextgen.loc[k,[1,2,3,4,5]]= mutation(list(df_nextgen.loc[k,[1,2,3,4,5]].astype("int")))

    b = []
    for k in range(0, df_nextgen.shape[0]):
        lista = df_nextgen.columns[(df_nextgen == 1.0).iloc[k]].values.tolist()
        lista = [x-1 for x in lista]
        b.append(lista)
    
    # Cálculo dos custos associados aos cromossomas após mutação
    df_nextgen.loc["filho1","Custo"]=find_solution(b[2])
    df_nextgen.loc["filho2","Custo"]=find_solution(b[3])
    
    print(" Matriz de cromossomas após mutações")
    print(df_nextgen)
    print("--------------------------------")
    
    var = df_inicial2["Custo"].nlargest(2).index
    df_inicial2.loc[var[0],:]= df_nextgen.loc["filho1"]
    df_inicial2.loc[var[1],:]= df_nextgen.loc["filho2"]
    print("Substituição dos filhos na população inicial")
    print(df_inicial2)



--------------------------------
                 Pais           
Pai1: 90
Pai2: 72
--------------------------------
     Matriz após crossover      
    1  2  3  4  5  Custo
90  0  0  1  1  1   9180
72  0  1  1  1  0   9230
--------------------------------
          1    2    3    4    5   Custo
90      0.0  0.0  1.0  1.0  1.0  9180.0
72      0.0  1.0  1.0  1.0  0.0  9230.0
filho1  0.0  0.0  1.0  1.0  0.0  9110.0
filho2  0.0  1.0  1.0  1.0  1.0  9720.0
--------------------------------
Cromossoma Filho1(após crossover):
1           0.0
2           0.0
3           1.0
4           1.0
5           0.0
Custo    9110.0
Name: filho1, dtype: float64

--------------------------------
Cromossoma Filho2(após crossover):
1           0.0
2           1.0
3           1.0
4           1.0
5           1.0
Custo    9720.0
Name: filho2, dtype: float64

--------------------------------
 Matriz de cromossomas após mutações
        1  2  3  4  5   Custo
90      0  0  1  1  1  9180.0
72      0  1  1  1  0  9

filho2  0.0  1.0  0.0  1.0  1.0  9000.0
--------------------------------
Cromossoma Filho1(após crossover):
1           0.0
2           1.0
3           0.0
4           1.0
5           1.0
Custo    9000.0
Name: filho1, dtype: float64

--------------------------------
Cromossoma Filho2(após crossover):
1           0.0
2           1.0
3           0.0
4           1.0
5           1.0
Custo    9000.0
Name: filho2, dtype: float64

--------------------------------
 Matriz de cromossomas após mutações
        1  2  3  4  5   Custo
90      0  1  0  1  1  9000.0
3       0  1  0  1  1  9000.0
filho1  0  1  0  1  1  9000.0
filho2  0  1  0  1  1  9000.0
--------------------------------
Substituição dos filhos na população inicial
    1  2  3  4  5  Custo
0   0  1  0  1  1   9000
2   0  1  0  1  1   9000
3   0  1  0  1  1   9000
12  0  1  0  1  1   9000
13  0  1  0  1  1   9000
16  0  1  0  1  1   9000
72  0  1  0  1  1   9000
73  0  1  0  1  1   9000
76  0  1  0  1  1   9000
90  0  1  0  1  1   9000